gateway/src/apicast/policy/token_introspection/token_introspection.lua

```diff
@@ -0,0 +1,70 @@
+local policy = require('apicast.policy')
+local _M = policy.new('Token Introspection Policy')
+
+local cjson = require('cjson.safe')
+local http_authorization = require 'resty.http_authorization'
+local http_ng = require 'resty.http_ng'
+local user_agent = require 'apicast.user_agent'
+local resty_env = require('resty.env')
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  --- authorization for the token introspection endpoint.
+  -- https://tools.ietf.org/html/rfc7662#section-2.2
+  local credential = 'Basic ' .. ngx.encode_base64(table.concat({ self.config.client_id or '', self.config.client_secret or '' }, ':'))
+  self.introspection_url = config.introspection_url
+  self.http_client = http_ng.new{
+    backend = config.client,
+    options = {
+      headers = {
+        ['User-Agent'] = user_agent(),
+        ['Authorization'] = credential
+      },
+      ssl = { verify = resty_env.enabled('OPENSSL_VERIFY') }
+    }
+  }
+  return self
+end
+
+--- OAuth 2.0 Token Introspection defined in RFC7662.
+-- https://tools.ietf.org/html/rfc7662
+local function introspect_token(self, token)
+  --- Parameters for the token introspection endpoint.
+  -- https://tools.ietf.org/html/rfc7662#section-2.1
+  local res, err = self.http_client.post(self.introspection_url , { token = token, token_type_hint = 'access_token'})
+  if err then
+    ngx.log(ngx.WARN, 'token introspection error: ', err, ' url: ', self.introspection_url)
+    return { active = false }
+  end
+
+  if res.status == 200 then
+    local token_info, decode_err = cjson.decode(res.body)
+    if decode_err then
+      ngx.log(ngx.ERR, 'failed to parse token introspection response:', decode_err)
+      return { active = false }
+    end
+    return token_info
```

Comment:
This will return whatever the JSON was. So if the JSON is `null` this would crash a bit later. 

I think it should verify `type(token_info) == 'table'` and return the error otherwise instead of checking just the `decode_err`.

Reply:
In my opinion, checking the `decode_err` is enough. Because it must be an error when `token_info` is null.
So, `cjson.decode()` always return `error` and `decode_err` must not be null  when `token_info` JSON is `null`. 

I checked `cjson.decode("")`,`cjson.decode("{}")`, `cjson.decode(nil)` and `cjson.decode(ngx.null)` returned `decode_err`.

Reply:
If the response body is `null` then the JSON will be correctly decoded to `ngx.null` like:

```lua
cjson.decode('null')
```

That will later crash on trying to access `.active` (attempt to index a userdata value):

```lua
cjson.decode('null').active
```

The same can happen when a number is returned like:

```lua
cjson.decode('42').active
```

What I'm saying is `introspect_token` always has to return a table, because later it tries to get the key `.access` from it. And the only way to verify it always returns table is to check the type.
Because the response body is basically user input it can be any valid JSON, including just string "null" or numbers.

Reply:
Thank you for information. I missed 'null' and numbers.

Do you think it is good as following? This worked fine with test code.
```lua
if decode_err or not (type(token_info) == 'table') then
  ngx.log(ngx.ERR, 'failed to parse token introspection response:', decode_err)
  return { active = false }
end
```

Reply:
I think it could be easier to read as:

```lua
if type(token_info) == 'table' then
  return token_info
else
  ngx.log(ngx.ERR, 'failed to parse token introspection response:', decode_err or 'not a table')
  return { active = false }
end
```

Reply:
Thank you! I think this is nice.